File: 0904-fruit-into-baskets/0904-fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int i=0;
        int j=0;
        unordered_map<int,int>um;
        int length=0;
        while(j<fruits.size())
        {
            if(!um.count(fruits[j]))
            {
                um[fruits[j]] =1;

            }
            else
            {
                um[fruits[j]]++;
            }

            if(um.size()>2)
            {
                while(um.size()>2)
                {
                    um[fruits[i]]--;
                       
                     if (um[fruits[i]] == 0)
                     {
                         um.erase(fruits[i]);
                     }
                       
                    i++;
                }
    
            }
            length=max(length,j-i+1);
            j++;
        }
        return length;
    }
};

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        int i=0;
        int j=0;
        int res=-1;
        HashMap<Integer,Integer>map=new HashMap<>();
        while(i<fruits.length && j <fruits.length)
        {
            if(map.containsKey(fruits[j]))
            {
                map.put(fruits[j],map.get(fruits[j])+1);
            }
            else
            {
                 map.put(fruits[j],1);
            }
            if(map.size() >2)
            {
             
              while(map.size() > 2)
              {
                  map.put( fruits[i] , map.get(fruits[i])-1);
                  if(map.get(fruits[i]) == 0)
                  {
                    map.remove(fruits[i]);
                  }
                  i++;

              }
             
            }
             int len=j-i+1;
              if(res<len)
              {
                res=len;
              }
            j++;

        }
        if(res==-1)
        {
            return 0;
        }
        return res;
        
    }
}